doc/specs/Keyboard-Selection.md

```diff
@@ -0,0 +1,143 @@
+---
+author: Carlos Zamora @cazamor
+created on: 2019-08-30
+last updated: 2019-09-21
+issue id: #715
+---
+
+# Keyboard Selection
+
+## Abstract
+
+This spec describes a new set of keybindings that allows the user to create and update a selection without the use of a mouse or stylus.
+
+## Inspiration
+
+ConHost allows the user to edit an existing selection using the keyboard. Holding `Shift` allows the user to move the second selection anchor in accordance with the arrow keys. The selection anchor updates by one cell per key event, allowing the user to refine the selected region. However, mouse selection seems to be superior in that it provides access to the following special selection modes:
+- Double Click: word selection
+- Triple Click: line selection
+
+Mark mode allows the users to update the selection anchors sequentially, and create a block selection.
+
+
+## Solution Design
+
+The fundamental solution design for keyboard selection is that the responsibilities between the Terminal Control and Terminal Core must be very distinct. The Terminal Control is responsible for handling user interaction and directing the Terminal Core to update the selection. The Terminal Core will need to update the selection according to the preferences of the Terminal Control.
+
+### Fundamental Terminal Control Changes
+
+`TermControl::_KeyDownHandler()` is responsible for interpretting the key events. At the time of writing this spec, there are 3 cases handled in this order:
+- Alt+Gr events
+- Keybindings
+- Send Key Event (also clears selection)
+
+Conditionally effective keybindings can be introduced to update the selection accordingly. A well known conditionally effective keybinding is the `Copy` action where text is copied if a selection is active, otherwise the key event is marked as `handled = false`. See "UI/UX Design" --> "Keybindings" for more information regarding keybindings.
+
+### Fundamental Terminal Core Changes
+
+The Terminal Core will need to expose a `KeyboardSelection()` function that is called by the keybinding handler. The following parameters will need to be passed in:

```

Comment:
Sorry, why? Shouldn't keyboard selection in the `Terminal Control` just be a slow motion simulation of mouse selection? I don't think I understand why the Core should know anything special about a keyboard. As far as I can see, there should be some union of mouse and keyboard activity that can be represented in a single set of selection motion interfaces that are either commanded by a Mouse or Keyboard handler in the Control and adapted into the appropriate repeats of motions on the Core.

Reply:
For mouse selection, we take a `COORD` then update the selection accordingly (we really just set the selection anchor to that). We have no interest in a `COORD` for keyboard selection, we just need a direction and an expansion mode.

If we want to translate a `Direction` and a `SelectionExpansionMode` to a `COORD` in `TermControl`, we wouldn't be leveraging the functionality that's already in `TerminalCore`.

Arguably, I would need a third parameter to decide which selection anchor to move (start or end?). The implemented function would look something like this (you can imagine how adding the selection anchor param would look):

```c++
void Terminal::KeyboardSelection(Direction dir, SelectionExpansionMode mode)
{
    switch (mode)
    {
    case SelectionExpansionMode::Cell:
        _UpdateAnchorByCell(dir);
        break;
    case SelectionExpansionMode::Word:
        _UpdateAnchorByWord(dir);
        break;
    case SelectionExpansionMode::Line:
        _UpdateAnchorByLine(dir);
        break;
    case SelectionExpansionMode::Viewport:
        _UpdateAnchorByViewport(dir);
        break;
    default:
        // the provided selection expansion mode is not supported
        return;
    }
}
```

Reply:
I sat and thought about this for a bit. I was imagining that the mouse movement would be able to just say "move the start or end anchor" and how far/where it would need to go as well. But I suppose that would mean that the Control would have to keep some sort of previous-click-pixel position and difference that to translate into a direction to go through the interface to the core just to be converted back into a coordinate plane. Maybe it's not worth it.

Reply:
I suppose after reading this and thinking about it, you're right to not re-use the mouse selection bits here, though maybe it would be useful to have an `h4` in the spec just explaining why we're not using the same code for both. It makes sense, but intuitively, you'd think they should be the same.